stdlib/public/BackDeployConcurrency/TaskPool.cpp

```diff
@@ -0,0 +1,872 @@
+        return PoolStatus{status - waiting};
+      }
+
+      /// Pretty prints the status, as follows:
+      /// PoolStatus{ P:{pending tasks} W:{waiting tasks} {binary repr} }
+      std::string to_string() {
+        std::string str;
+        str.append("PoolStatus{ ");
+        str.append("C:"); // cancelled
+        str.append(isCancelled() ? "y " : "n ");
+        str.append("W:"); // has waiting task
+        str.append(hasWaitingTask() ? "y " : "n ");
+        str.append(" P:"); // pending
+        str.append(std::to_string(pendingTasks()));
+        str.append(" " + std::bitset<64>(status).to_string());
+        str.append(" }");
+        return str;
+      }
+
+      /// Initially there are no waiting and no pending tasks.
+      static const PoolStatus initial() {
+        return PoolStatus{0};
+      };
+    };
+
+    // =============================================================================
+    // ==== Checks -----------------------------------------------------------------
+
+    static void _taskPool_reportPendingTaskOverflow(TaskPoolImpl *pool, uint64_t pendingCount) {
+
+      char *message;
+      swift_asprintf(
+          &message,
+          "error: task-pool: detected pending task overflow, in task pool %p! Pending task count: %llu",
+          pool, pendingCount);
+
+      if (_swift_shouldReportFatalErrorsToDebugger()) {
+        RuntimeErrorDetails details = {
+            .version = RuntimeErrorDetails::currentVersion,
+            .errorType = "task-pool-violation",
+            .currentStackDescription = "Task-pool exceeded supported pending task count",
+            .framesToSkip = 1,
+        };
+        _swift_reportToDebugger(RuntimeErrorFlagFatal, message, &details);
+      }
+
+#if defined(_WIN32)
+      #define STDERR_FILENO 2
+  _write(STDERR_FILENO, message, strlen(message));
+#else
+      write(STDERR_FILENO, message, strlen(message));
+#endif
+#if defined(__APPLE__)
+      asl_log(nullptr, nullptr, ASL_LEVEL_ERR, "%s", message);
+#elif defined(__ANDROID__)
+      __android_log_print(ANDROID_LOG_FATAL, "SwiftRuntime", "%s", message);
+#endif
+
+      free(message);
+      abort();
+    }
+
+  private:
+#if SWIFT_STDLIB_SINGLE_THREADED_CONCURRENCY || SWIFT_CONCURRENCY_TASK_TO_THREAD_MODEL
+    // Synchronization is simple here. In a single threaded mode, all swift tasks
+  // run on a single thread so no coordination is needed. In a task-to-thread
+  // model, only the parent task which created the task pool can
+  //
+  //   (a) add child tasks to a pool
+  //   (b) run the child tasks
+  //
+  // So we shouldn't need to worry about coordinating between child tasks and
+  // parents in a task pool
+  void lock() const {}
+  void unlock() const {}
+#else
+    // TODO: move to lockless via the status atomic (make readyQueue an mpsc_queue_t<ReadyQueueItem>)
+    mutable std::mutex mutex_;
+
+    void lock() const { mutex_.lock(); }
+    void unlock() const { mutex_.unlock(); }
+#endif
```

Comment:
TODO: we _definitely_ can remove this lock; so will do so once we have agreed on exact semantics of the pool.